Zend/zend_hash.h

```diff
@@ -1023,114 +1027,112 @@ static zend_always_inline void *zend_hash_get_current_data_ptr_ex(HashTable *ht,
 
 #define ZEND_HASH_FOREACH_NUM_KEY(ht, _h) \
 	ZEND_HASH_FOREACH(ht, 0); \
-	_h = _p->h;
+	ZEND_ASSERT(Z_TYPE(_p->key) == IS_LONG); \
+	_h = Z_LVAL(_p->key);
 
 #define ZEND_HASH_REVERSE_FOREACH_NUM_KEY(ht, _h) \
 	ZEND_HASH_REVERSE_FOREACH(ht, 0); \
-	_h = _p->h;
+	ZEND_ASSERT(Z_TYPE(_p->key) == IS_LONG); \
+	_h = Z_LVAL(_p->key);
 
 #define ZEND_HASH_FOREACH_STR_KEY(ht, _key) \
 	ZEND_HASH_FOREACH(ht, 0); \
-	_key = _p->key;
+	_key = EXPECTED(Z_TYPE(_p->key) == IS_STRING) ? Z_STR(_p->key) : NULL;
 
 #define ZEND_HASH_REVERSE_FOREACH_STR_KEY(ht, _key) \
 	ZEND_HASH_REVERSE_FOREACH(ht, 0); \
-	_key = _p->key;
+	_key = EXPECTED(Z_TYPE(_p->key) == IS_STRING) ? Z_STR(_p->key) : NULL;
 
-#define ZEND_HASH_FOREACH_KEY(ht, _h, _key) \
+#define ZEND_HASH_FOREACH_ZKEY(ht, _key) \
 	ZEND_HASH_FOREACH(ht, 0); \
-	_h = _p->h; \
-	_key = _p->key;
+	_key = &_p->key;
 
-#define ZEND_HASH_REVERSE_FOREACH_KEY(ht, _h, _key) \
+#define ZEND_HASH_REVERSE_FOREACH_ZKEY(ht, _key) \
 	ZEND_HASH_REVERSE_FOREACH(ht, 0); \
-	_h = _p->h; \
-	_key = _p->key;
+	_key = &_p->key;
 
 #define ZEND_HASH_FOREACH_NUM_KEY_VAL(ht, _h, _val) \
 	ZEND_HASH_FOREACH(ht, 0); \
-	_h = _p->h; \
+	ZEND_ASSERT(Z_TYPE(_p->key) == IS_LONG); \
+	_h = Z_LVAL(_p->key); \
 	_val = _z;
 
 #define ZEND_HASH_REVERSE_FOREACH_NUM_KEY_VAL(ht, _h, _val) \
 	ZEND_HASH_REVERSE_FOREACH(ht, 0); \
-	_h = _p->h; \
+	ZEND_ASSERT(Z_TYPE(_p->key) == IS_LONG); \
+	_h = Z_LVAL(_p->key); \
 	_val = _z;
 
 #define ZEND_HASH_FOREACH_STR_KEY_VAL(ht, _key, _val) \
 	ZEND_HASH_FOREACH(ht, 0); \
-	_key = _p->key; \
+	_key = EXPECTED(Z_TYPE(_p->key) == IS_STRING) ? Z_STR(_p->key) : NULL; \
 	_val = _z;
 
 #define ZEND_HASH_REVERSE_FOREACH_STR_KEY_VAL(ht, _key, _val) \
 	ZEND_HASH_REVERSE_FOREACH(ht, 0); \
-	_key = _p->key; \
+	_key = EXPECTED(Z_TYPE(_p->key) == IS_STRING) ? Z_STR(_p->key) : NULL; \
 	_val = _z;
 
-#define ZEND_HASH_FOREACH_KEY_VAL(ht, _h, _key, _val) \
+#define ZEND_HASH_FOREACH_ZKEY_VAL(ht, _key, _val) \
```

Comment:
The ZKEY macros and functions are already intended to be polyfilled on older PHP versions. I've found that in most cases it's more convenient to keep the key as a zval rather than unpacking it, as most usages just pass through the key in some manner.